### PR TITLE
Colab Resolver

### PR DESCRIPTION
**Pull request by mayankmalik-colab**
b/303240466

### CODE REVIEW

**Review by rosbo — APPROVED** [review]
Great work Mayank! Just a few minor comments.

Can you run the linter too:

```
hatch run lint:fmt
```

https://github.com/Kaggle/kagglehub?tab=readme-ov-file#lint--format

File: src/kagglehub/clients.py [hunk 1]
```diff
@@ -209,3 +211,43 @@ def post(
...
+
+    def __init__(self):
+        self.endpoint = os.getenv(self.TBE_RUNTIME_ADDR_ENV_VAR_NAME)
+        if self.endpoint is None:
+            msg = (
+                f"The {self.TBE_RUNTIME_ADDR_ENV_VAR_NAME} should be set."
+            )
+            raise ColabEnvironmentError(msg)
+
+        self.credentials = get_kaggle_credentials()
```
[the remark this answers is not in the document]
Comment (mayankmalik-colab) [reply src/kagglehub/clients.py:229]: They are used in `_get_http_basic_auth` and then in the post request. So, I am leaving this for now.

File: tests/test_colab_cache_model_download.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    def do_HEAD(self):  # noqa: N802
+        self.send_response(200)
+
+    def do_POST(self):  # noqa: N802
+        content_length = int(self.headers["Content-Length"])
+        request = json.loads(self.rfile.read(content_length))
+        version = LATEST_MODEL_VERSION
+        if "version" in request:
+            version = request["version"]
+        print("request ", request)
```
[the remark this answers is not in the document]
Comment (mayankmalik-colab) [reply tests/test_colab_cache_model_download.py:32]: Oops. My bad. Removed in the next commit.

File: src/kagglehub/clients.py [hunk 3]
```diff
@@ -209,3 +211,43 @@ def post(
                 msg = "'result' field missing from response"
                 raise BackendError(msg)
             return json_response["result"]
+
+
+class ColabClient:
+    IS_SUPPORTED_PATH = "/kagglehub/models/is_supported"
+    MOUNT_PATH = "/kagglehub/models/mount"
+    TBE_RUNTIME_ADDR_ENV_VAR_NAME = "TBE_RUNTIME_ADDR"
```
[the remark this answers is not in the document]
Comment (mayankmalik-colab) [reply src/kagglehub/clients.py:219]: Done in the new commit.

File: src/kagglehub/config.py [hunk 4]
```diff
@@ -104,6 +106,10 @@ def get_log_verbosity() -> int:
     return DEFAULT_LOG_LEVEL
 
 
+def is_colab_cache_disabled() -> bool:
```
[the remark this answers is not in the document]
Comment (mayankmalik-colab) [reply src/kagglehub/config.py:109]: Done.

File: tests/test_colab_cache_model_download.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+            self.wfile.write(
+                bytes(
+                    json.dumps(
+                        {
+                            "slug": slug,
+                        }
+                    ),
+                    "utf-8",
+                )
+            )
```
[the remark this answers is not in the document]
Comment (mayankmalik-colab) [reply tests/test_colab_cache_model_download.py:66]: Ahh. Makes sense. Done.

File: tests/test_colab_cache_model_download.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        with create_test_server_colab(ColabTBERuntimeHandler):
+            with self.assertRaises(ValueError):
+                kagglehub.model_download(VERSIONED_MODEL_HANDLE, "missing.txt")
+
+    def test_unversioned_model_download_with_missing_file_raises(self):
+        with create_test_server_colab(ColabTBERuntimeHandler):
+            with self.assertRaises(ValueError):
+                kagglehub.model_download(UNVERSIONED_MODEL_HANDLE, "missing.txt")
+
+    def test_colab_resolver_skipped(self):
```
[the remark this answers is not in the document]
Comment (mayankmalik-colab) [reply tests/test_colab_cache_model_download.py]: Sure. Just added.

**Comment by rosbo** [comment]
Here is the linter failure:

```
Step #3 - "lint": src/kagglehub/clients.py:233: error: Missing return statement  [return]
Step #3 - "lint": src/kagglehub/colab_cache_resolver.py:18: error: Signature of "is_supported" incompatible with supertype "Resolver"  [override]
Step #3 - "lint": src/kagglehub/colab_cache_resolver.py:18: note:      Superclass:
Step #3 - "lint": src/kagglehub/colab_cache_resolver.py:18: note:          def is_supported(self, handle: ModelHandle, path: str | None = ...) -> Any
Step #3 - "lint": src/kagglehub/colab_cache_resolver.py:18: note:      Subclass:
Step #3 - "lint": src/kagglehub/colab_cache_resolver.py:18: note:          def is_supported(self, h: ModelHandle, *_: Any, **__: Any) -> bool
Step #3 - "lint": src/kagglehub/colab_cache_resolver.py:31: error: Incompatible types in assignment (expression has type "int | None", target has type "str")  [assignment]
Step #3 - "lint": src/kagglehub/colab_cache_resolver.py:53: error: Incompatible types in assignment (expression has type "int | None", target has type "str")  [assignment]
```

You should be seeing the same error locally when running `hatch run lint:all`. Let me know if that is not the case.

**Comment by rosbo** [comment]
Yeah! All tests are passing now. Congrats @mayankmalik-colab on your first PR to `kagglehub` and thank you :)

**Pull request merged**